SPECS/tdnf/virtual-repo-snapshot.patch

```diff
@@ -0,0 +1,906 @@
++                pszStartElementBuffer = malloc(sizeof(char) * nStartElementBufferSize);
++                if (!pszStartElementBuffer)
++                {
++                    dwError = ERROR_TDNF_TIME_FILTER_MEMORY;
++                    BAIL_ON_TDNF_TIME_FILTER_ERROR(dwError);
++                }
++                int nLenToAdd = addElementStartToBuffer(pszStartElementBuffer, &nStartElementBufferSize, name, attrs);
++                strcat(pszStartElementBuffer, szNewLineBuffer);
++                nLenToAdd += strlen(szNewLineBuffer); //+1 if newLine character present
++
++                checkAndResizeBuffer(&(pTracking->pszElementBuffer), &(pTracking->nBufferMaxLen), nLenToAdd);
++                strcat(pTracking->pszElementBuffer, pszStartElementBuffer);
++            }
++        }
++    }
++    else
++    { // not in a package or parsing a new package
++        fprintf(pTracking->pbOutfile, "%s", szNewLineBuffer);
++        // output line
++        printElementStartToFile(pTracking->pbOutfile, name, attrs);
++    }
++cleanup:
++    if (pszStartElementBuffer)
++    {
++        free(pszStartElementBuffer);
++    }
++    return;
++error:
++    goto cleanup;
++}
++
++void TDNFFilterEndElement(void * userData, const char * name) {
++    uint32_t dwError = 0;
++    char * pszElementBuffer = NULL;
++    // load tracking data
++    TDNFFilterData * pTracking = (TDNFFilterData *)userData;
++
++    // decrement depth
++    pTracking->nDepth -= 1;
++    pTracking->nPrevElement = 2;
++
++    if (!pTracking->nInPackage || pTracking->nPrintPackage)
++    {
++        // print end element to file
++        printElementEndToFile(pTracking->pbOutfile, name);
++
++    } 
++    else if (pTracking->nInPackage && !pTracking->nTimeFound)
++    {
++        int nEndElementBufferLen = 500;
++        pszElementBuffer = malloc(sizeof(char) * nEndElementBufferLen);
++        if (!pszElementBuffer)
++        {
++            dwError = ERROR_TDNF_TIME_FILTER_MEMORY;
++            BAIL_ON_TDNF_TIME_FILTER_ERROR(dwError);
++        }
++        // add end element to buffer
++        int nEndElementLen = addElementEndToBuffer(pszElementBuffer, &nEndElementBufferLen, name);
++
++        checkAndResizeBuffer(&(pTracking->pszElementBuffer), &(pTracking->nBufferMaxLen), nEndElementLen);
++        strcat(pTracking->pszElementBuffer, pszElementBuffer);
++
++    } // else do nothing
++
++    if (strcmp(name, "package") == 0)
++    { // on end package, reset tracking function
++        // reset userData
++        pTracking->nBufferLen = 0;
++        bzero(pTracking->pszElementBuffer, pTracking->nBufferMaxLen);
++        pTracking->nInPackage = 0;
++        pTracking->nPrintPackage = 0;
++        pTracking->nTimeFound = 0;
++    }
++cleanup:
++    if (pszElementBuffer)
++    {
++        free(pszElementBuffer);
++    }
++    return;
++error:
++    goto cleanup;
++}
++
++void TDNFFilterCharDataHandler(void * userData, const char * content, int length) {
```

Comment:
Why take void* here if the data type is known to be TDNFFilterData*

Reply:
This is the function handle/format used for expat's character parsing. That is why the first thing done in each handler is to cast it to TDNFFilterData. Unfortunately, I don't think we can change this easily...